File: algorithms/javascript/search-a-2d-matrix-ii/search-a-2d-matrix-ii.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const rowLength = matrix.length;
    const columnLength = matrix[0].length;

    const recSearchMatrix = (rowStart, rowEnd, columnStart, columnEnd) => {
        if (rowStart >= rowEnd || columnStart >= columnEnd) {
            return false;
        }
        let rowPivot = rowStart;
        const columnPivot = Math.floor((columnStart + columnEnd) / 2);
        while (rowPivot < rowEnd) {
            if (matrix[rowPivot][columnPivot] === target) {
                return true;
            } else if (matrix[rowPivot][columnPivot] > target) {
                return recSearchMatrix(rowStart, rowEnd, columnStart, columnPivot); 
            } else if (rowPivot + 1 < rowEnd && matrix[rowPivot][columnPivot] < target && matrix[rowPivot + 1][columnPivot] > target) {
                return recSearchMatrix(rowStart, rowPivot + 1, columnPivot + 1, columnEnd) || recSearchMatrix(rowPivot, rowEnd, columnStart, columnPivot);
            }
            rowPivot++;
        }
        return recSearchMatrix(rowStart, rowEnd, columnPivot + 1, columnEnd);
    }

    return recSearchMatrix(0, rowLength, 0, columnLength);
};

File: algorithms/go/search-a-2d-matrix-ii/search-a-2d-matrix-ii.go
func searchMatrix(matrix [][]int, target int) bool {
	return recSearchMatrix(matrix, target, 0, len(matrix), 0, len(matrix[0]))
}

func recSearchMatrix(matrix [][]int, target int, rowStart, rowEnd, colStart, colEnd int) bool {
	if rowStart >= rowEnd || colStart >= colEnd {
		return false
	}
	rowPivot, colPivot := rowStart, (colStart+colEnd)/2
	for rowPivot < rowEnd {
		if matrix[rowPivot][colPivot] == target {
			return true
		} else if matrix[rowPivot][colPivot] > target {
			return recSearchMatrix(matrix, target, rowStart, rowEnd, colStart, colPivot)
		} else if rowPivot+1 < rowEnd && matrix[rowPivot+1][colPivot] > target && matrix[rowPivot][colPivot] < target {
			return recSearchMatrix(matrix, target, rowStart, rowPivot+1, colPivot+1, colEnd) || recSearchMatrix(matrix, target, rowPivot+1, rowEnd, colStart, colPivot)
		}
		rowPivot++
	}
	return recSearchMatrix(matrix, target, rowStart, rowEnd, colPivot+1, colEnd)
}